app/src/block_hash.rs

```diff
@@ -1,37 +1,226 @@
+    /// Block number of the latest block stored for validation purposes
+    latest_block: u64,
 }
 
-impl<'a, S, B> StorageBasedProvider<'a, S, B>
-where
-    B: BlockQueries<Storage = S>,
-{
-    pub fn new(storage: &'a S, block_query: &'a B) -> Self {
+impl Default for BlockHashRingBuffer {
+    fn default() -> Self {
         Self {
-            storage,
-            block_query,
+            entries: VecDeque::with_capacity(BLOCKHASH_HISTORY_SIZE),
+            latest_block: 0,
         }
     }
 }
 
-// TODO: looking up the entire block just to get its hash is inefficient;
-// especially since we do not need the whole history, only the most recent 256
-// block hashes. This implementation should be replaced with something better.
-impl<S, B> BlockHashLookup for StorageBasedProvider<'_, S, B>
-where
-    B: BlockQueries<Storage = S>,
-{
+impl BlockHashRingBuffer {
+    pub fn push(&mut self, block_number: u64, block_hash: B256) {
+        // Ensure blocks are added in sequence (detect potential reorgs)
+        if block_number > 0 && self.latest_block > 0 && block_number != self.latest_block + 1 {
+            eprintln!(
+                "WARN: Block hash buffer - expected block {}, got {}",
+                self.latest_block + 1,
+                block_number
+            );
+        }
+
+        self.entries.push_back(BlockHashEntry {
+            number: block_number,
+            hash: block_hash,
+        });
+
+        while self.entries.len() > BLOCKHASH_HISTORY_SIZE {
+            self.entries.pop_front();
+        }
+
+        self.latest_block = block_number;
+    }
+
+    // TODO: make this a builder method?
+    pub fn initialize_from_storage<S, B>(&mut self, storage: &S, block_query: &B, latest_block: u64)
+    where
+        B: BlockQueries<Storage = S>,
+    {
+        self.entries.clear();
+        self.latest_block = 0;
+
+        let start_block = if latest_block >= BLOCKHASH_HISTORY_SIZE as u64 {
+            latest_block - BLOCKHASH_HISTORY_SIZE as u64 + 1
+        } else {
+            0
+        };
+
+        for block_num in start_block..=latest_block {
+            if let Ok(Some(block)) = block_query.by_height(storage, block_num, false) {
+                self.push(block_num, block.0.header.hash);
+            }
+        }
+    }
+}
+
+impl BlockHashLookup for BlockHashRingBuffer {
+    fn hash_by_number(&self, block_number: u64) -> Option<B256> {
+        if block_number > self.latest_block {
+            // Future block
+            return None;
+        }
+
+        let blocks_ago = self.latest_block - block_number;
+        if blocks_ago > BLOCKHASH_HISTORY_SIZE as u64 {
+            // Too old
+            return None;
+        }
+
+        for entry in self.entries.iter().rev() {
+            if entry.number == block_number {
+                return Some(entry.hash);
+            }
+            // Since we're going backwards chronologically, if we've passed the target block, it's not there
```

Comment:
Can this happen if blocks are always in sequence?

Reply:
This loops back to that `eprintln!` message [above](https://github.com/UmiNetwork/op-move/pull/375#discussion_r2117159789), as that's the only case where it could happen in a running application. Though considering that it's meant to be initialized from storage at startup, I can also see a scenario where there's some uncaught data corruption happening that jumbled up this order.